relayer/workers/ethrelayer/main.go

```diff
@@ -0,0 +1,160 @@
+package ethrelayer
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/sirupsen/logrus"
+	"github.com/snowfork/go-substrate-rpc-client/v2/types"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain/ethereum"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain/parachain"
+	"github.com/snowfork/polkadot-ethereum/relayer/crypto/sr25519"
+)
+
+type Worker struct {
+	parachainConfig  *parachain.Config
+	ethereumConfig   *ethereum.Config
+	parachainConn    *parachain.Connection
+	ethereumConn     *ethereum.Connection
+	ethereumListener *EthereumListener
+	parachainWriter  *ParachainWriter
+	log              *logrus.Entry
+}
+
+const Name = "eth-relayer"
+
+func NewWorker(ethereumConfig *ethereum.Config, parachainConfig *parachain.Config) (*Worker, error) {
+	log := logrus.WithField("worker", Name)
+
+	log.Info("Creating worker")
+
+	ethereumConn := ethereum.NewConnection(ethereumConfig.Endpoint, nil, log)
+
+	// Generate keypair from secret
+	parachainKeypair, err := sr25519.NewKeypairFromSeed(parachainConfig.PrivateKey, "")
+	if err != nil {
+		return nil, err
+	}
+	parachainConn := parachain.NewConnection(parachainConfig.Endpoint, parachainKeypair.AsKeyringPair(), log)
+
+	// channel for messages from ethereum
+	ethMessages := make(chan []chain.Message, 1)
+	// channel for headers from ethereum (it's a blocking channel so that we
+	// can guarantee that a header is forwarded before we send dependent messages)
```

Comment:
How exactly do you mean this? The guarantee only holds if there is one messages array received before the header, not if there are multiple arrays of messages received.